cusub_perception/detection_tree/src/detection_tree_node.cpp

```diff
@@ -0,0 +1,360 @@
+        int ret = transformBearingToOdom(odom_cam_pose, bearing_vec, image_header);
+        if (ret) continue; // failed transform
+        debug_dv_pose_pub.publish(odom_cam_pose);
+
+        // Extract roll, pitch, yaw in odom frame
+        double roll_odom, pitch_odom, yaw_odom;
+        tf2::Quaternion odom_quat;
+        tf2::fromMsg(odom_cam_pose.pose.orientation, odom_quat);
+        tf2::Matrix3x3 m(odom_quat);
+        m.getRPY(roll_odom, pitch_odom, yaw_odom);
+
+        // Create Dvector
+        detection_tree::Dvector* dv = new detection_tree::Dvector; // we'll be storing it globally
+        dv->sub_pt = odom_cam_pose.pose.position;
+        dv->azimuth = yaw_odom;
+        dv->elevation = pitch_odom;
+        dv->camera_header = image_header;
+        dv->class_id = box.Class;
+        dv->probability = box.probability;
+        dv->magnitude = (box.xmax - box.xmin) * (box.ymax - box.ymin);
+        dv->dvector_num = dvector_num;
+        dvector_num++;
+        dv_list.push_back(dv);
+    }
+    map<detection_tree::Dvector*, int> dv_dobj_map;
+    associateDvectors(dv_list, dv_dobj_map);
+
+    // Add dvectors to dobjects or create a new dobject
+    // Publish dvectors
+    for ( auto it = dv_dobj_map.begin(); it != dv_dobj_map.end(); it++ )
+    {
+        detection_tree::Dvector* dv = it->first;
+        int dobj_num = it->second;
+        // Create new dobject
+        if( dobj_num == DOBJECT_NOT_FOUND)
+        {
+            dobj_num = createDobject(dv);
+        } 
+        else {
+            dobject_list[dobj_num]->dvector_list.push_back(dv);
+        }
+        dv->dobject_num = dobj_num;
+        dvector_pub.publish(*dv);
+    }
+}
+
+/*  @brief creates a new dobject
+    @param first dvector
+    @return dobject number
+*/
+int DetectionTree::createDobject(detection_tree::Dvector* dv)
+{   
+    string s = string("creating new dobj for ") + DETECTION_TREE_COLOR_START + dv->class_id + DETECTION_TREE_END;
+    det_print(s);
+    Dobject* dobj = new Dobject;
+    dobj->num = dobject_list.size();
+    dobj->class_id = dv->class_id;
+    dobj->pose = geometry_msgs::Pose();
+    dobj->dvector_list.push_back(dv);
+    dobject_list.push_back(dobj);
+    return dobj->num;
+}
+
+void DetectionTree::averageBearing(vector<detection_tree::Dvector*>& dvs, double& average_az, double& average_elev, double& average_mag)
+{
+    average_az = 0.0;
+    average_elev = 0.0;
+    average_mag = 0.0;
+    for ( auto dv : dvs )
+    {
+        average_az += dv->azimuth;
+        average_elev += dv->elevation;
+        average_mag += dv->magnitude;
+    }
+    average_az /= dvs.size();
+    average_elev /= dvs.size();
+    average_mag /= dvs.size();
+}
+
+void DetectionTree::assignDobjScores(std::vector<detection_tree::Dvector*>& dv_list, map<detection_tree::Dvector*, map<int, double>*>& dvs_scored)
+{
+    for ( auto dv : dv_list )
+    {
+        map<int, double>* dobj_scores = new map<int, double>();
```

Comment:
Is this pointer ever cleaned up? Could be a memory leak

Reply:
Yes in line 291 of that file all of the maps created get freed

Reply:
Ah, I didn't notice this map is put into another map and returned